### DISCUSSION TITLE
Importer doesn't allow data[] values not in getColumns array

### DISCUSSION THREAD

**Discussion by djamesfar** [filamentphp/filament#11046] [Help] [answered]
### Package

Widgets Importer

### Package Version

v3.1

### How can we help you?

I have an importer that uses a csv file with some, but not all of the fields in the table.  The rest of the required table fields need to get populated during the import.  
It seems that any fields not defined with `ImportColumn::make()` in `getColumns()` cannot be added to the $data array.  I have tried using the lifecycle hooks like `beforeFill()`, `afterFill()` and `beforeSave()`, for example:
```
protected function afterFill(): void
{
     // get the term
     $term = Term::find(getCurrentTermId());
     $this->data['license_expires'] = $term->end_date->addWeek();
     // other assignments...
}
```
This does not add anything to the data array.  I tried the same thing in `beforeSave()`, no joy.

I can add some values to fields in `$this->record` in the `afterSave()` method, but obviously this won't work for required fields as the record can't be saved without them.
My solution was to add the columns using `ImportColumn::make()`, and then  `->label('ignore')` so the user understands they don't need to map the field to a column in the import csv file.  But this shows a long list of unnecessary mappings that are confusing to the user.
This allows me to populate the column in a callback using `->fillRecordUsing()`.  

I cannot find a function to hide the column from the user, similar to the table resource's 
`->toggleable(isToggledHiddenByDefault: true)`.

Am I missing some way to either add elements to the `$data` array in the lifecycle hooks, or else to hide columns assigned with `ImportColumn::make()` so they don't show up in the user facing header mapping modal?

**Comment by rhysflook** [comment, marked as answer]
I was able to add data using the lifecycle hooks. But if they are giving you issues you could try overriding the `Importer` classes `saveRecord` method.

```
    public function saveRecord(): void
    {
        $this->record->password = \Hash::make(PasswordGenerator::generatePassword());
        $this->record->save();
    }
```

**Reply by rhysflook** [reply]
But come to think of it I did have issues with the lifecycle hooks at first. I think the data may be cached so try clearing your caches as well.

**Reply by djamesfar** [reply]
@rhysflook , you were able to add data in columns that were not in your importColumn::make()?  Which hook did you use?

**Comment by djamesfar** [comment]
Okay, I confirmed that when you add data for fields not in the import columns array, they show up in the data array in beforeFill, but they _**do not get saved in the record**_!  This seems like a bug.
I dumped the $data array in beforeSave, and all the non-imported fields I had populated show up in data[].  However, afterSave shows the only fields that got written to the record were the ones in the importer array.
@rhysflook's suggestion of overriding saveRecord() to add the additional column data did the trick.
